### PR TITLE
added test program for priority inversion problem

### PR DESCRIPTION
**Pull request by geith**
added simple test program for priority inversion problem as discussed in #7365

### CODE REVIEW

File: tests/thread_priority_inversion/Makefile [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+CFLAGS += -DDEVELHELP
+
+# Change this to 0 show compiler invocation lines by default:
+QUIET ?= 1
+
+# Modules to include:
+USEMODULE += shell
+USEMODULE += shell_commands
+USEMODULE += ps
+USEMODULE += xtimer
```
Comment (smlng) [inline tests/thread_priority_inversion/Makefile]: only `xtimer` is needed for your test, the 3 modules above can safely be removed.

File: tests/thread_priority_inversion/Makefile [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# name of your application
+APPLICATION = Scheduling-Test
```
Comment (smlng) [inline tests/thread_priority_inversion/Makefile]: this should be the same as the directory name -> `thread_priority_inversion`

File: tests/thread_priority_inversion/main.c [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+char stack_high[THREAD_STACKSIZE_MAIN];
+char stack_mid[THREAD_STACKSIZE_MAIN];
+char stack_low[THREAD_STACKSIZE_MAIN];
+
+void *t_low_handler(void *arg)
+{
+    /* starting working loop immediately */
+	while(1){
+		printf("t_low: allocating resource...\n");
```
Comment (smlng) [inline tests/thread_priority_inversion/main.c]: replace `printf` with`puts` here and below. `puts` can output static strings only, but is less expensive than `printf`.

Comment (gebart) [inline tests/thread_priority_inversion/main.c]: IIRC GCC should replace printf by puts automatically when optimizing if the string is static and ends in a newline.

**Comment by geith** [comment]
requested changes done.

File: tests/thread_priority_inversion/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
+# thread_priority_inversion test application
+
+This application uses three threads for demonstrating the
+priority inversion problem. In theory, the highest priority thread (**t_high**)
+should periodically and produce some output:
```
Comment (smlng) [inline tests/thread_priority_inversion/README.md]: missing word? `... should [be scheduled] periodically ...`

File: tests/thread_priority_inversion/README.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+```
+2017-07-17 17:00:29,337 - INFO # t_high: got resource.
+...
+2017-07-17 17:00:30,343 - INFO # t_high: freeing resource...
+```
+During this phase of 1s, **t_high** lockes the mutex **res_mtx** which
+represents a shared ressource. After unlocking the mutex, **t_high** waits 1s
+before restaring tis cycle again.
```
Comment (smlng) [inline tests/thread_priority_inversion/README.md]: typo: /tis/its/

File: tests/thread_priority_inversion/README.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+A low priority thread **t_low** is doing the same. Since both threads sharing
+the same resource **res_mtx**, they have to wait for wach other until the mutex
```
Comment (smlng) [inline tests/thread_priority_inversion/README.md]: typo: /wach/each/

File: tests/thread_priority_inversion/README.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+2017-07-17 17:00:30,343 - INFO # t_high: freeing resource...
+2017-07-17 17:00:30,344 - INFO # t_high: freed resource.
+2017-07-17 17:00:30,345 - INFO # t_low: allocating resource...
+2017-07-17 17:00:30,346 - INFO # t_low: got resource.
+...
+```
+
+After 3s, a third thread with medium priority (**t_mid**) is started. This
+thread does not touch **res_mtx**, but it runs an infinite loop without leaving
+come CPU time to lower priority tasks. This prevents **t_low** from freeing the
```
Comment (smlng) [inline tests/thread_priority_inversion/README.md]: typo: /come/some/

File: tests/thread_priority_inversion/main.c [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+mutex_t res_mtx;
+
+char stack_high[THREAD_STACKSIZE_MAIN];
+char stack_mid[THREAD_STACKSIZE_MAIN];
+char stack_low[THREAD_STACKSIZE_MAIN];
+
+void *t_low_handler(void *arg)
+{
+    /* starting working loop immediately */
+	while(1){
```
Comment (smlng) [inline tests/thread_priority_inversion/main.c]: indention seems off

File: tests/thread_priority_inversion/main.c [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+		puts("t_low: freed resource.");
+		xtimer_sleep(1);
+	}
+	return NULL;
+}
+
+void *t_mid_handler(void *arg)
+{
+    /* starting working loop after 3s */
+	xtimer_sleep(3);
```
Comment (smlng) [inline tests/thread_priority_inversion/main.c]: indention seem off

**Comment by smlng** [comment]
also consider squashing

**Comment by geith** [comment]
requested changes done...

**Comment by smlng** [comment]
Looks better now, however Murdock isn't happy - some minor issues with unused variables (add `(void)arg;` for instance) and a whitespace error in Makefile. Please fix, Pre-ACK.

**Comment by geith** [comment]
unused variables and makefile issues fixed...

**Comment by smlng** [comment]
we are nearly there, you need to add the following directive to the Makefile of your test
```
BOARD_INSUFFICIENT_MEMORY := nucleo32-f031 nucleo32-f042
```

that should fix last problem!

**Comment by geith** [comment]
done.

File: tests/thread_priority_inversion/main.c [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+char stack_high[THREAD_STACKSIZE_MAIN];
+char stack_mid[THREAD_STACKSIZE_MAIN];
+char stack_low[THREAD_STACKSIZE_MAIN];
+
+void *t_low_handler(void *arg)
+{
+    (void) arg;
+
+    /* starting working loop immediately */
+    while(1){
```
Comment (smlng) [inline tests/thread_priority_inversion/main.c]: coding style here and 2x below: add missing spaces around parentheses, should be `while(1) {`

File: tests/thread_priority_inversion/main.c [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+void *t_mid_handler(void *arg)
+{
+    (void) arg;
+
+    /* starting working loop after 3s */
+    xtimer_sleep(3);
+
+    puts("t_mid: doing some stupid stuff...");
+    while(1){
```
Comment (smlng) [inline tests/thread_priority_inversion/main.c]: here

File: tests/thread_priority_inversion/main.c [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    }
+}
+
+void *t_high_handler(void *arg)
+{
+    (void) arg;
+
+    /* starting working loop after 500 ms */
+    xtimer_usleep(500);
+    while(1){
```
Comment (smlng) [inline tests/thread_priority_inversion/main.c]: abd here, too

File: tests/thread_priority_inversion/main.c [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+        thread_yield_higher();
+    }
+}
+
+void *t_high_handler(void *arg)
+{
+    (void) arg;
+
+    /* starting working loop after 500 ms */
+    xtimer_usleep(500);
```
Comment (smlng) [inline tests/thread_priority_inversion/main.c]: this actually only 500 microseconds, if you want milliseconds change to: `500U * US_PER_MS`

File: tests/thread_priority_inversion/main.c [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+#include <string.h>
+#include <stdint.h>
+
+#include "thread.h"
+#include "mutex.h"
+#include "xtimer.h"
+
+mutex_t res_mtx;
+
+char stack_high[THREAD_STACKSIZE_MAIN];
```
Comment (smlng) [inline tests/thread_priority_inversion/main.c]: IMHO `THREAD_STACKSIZE_DEFAULT` will be sufficient, too and will lower required memory considerably.

**Comment by geith** [comment]
changes done; compiles now for the nucleo32-f042 too (but still not for the nucleo32-f031)

**Comment by smlng** [comment]
tested on macOS and PhyNode works (or actually for now: fails) as expected 😉

**Comment by smlng** [comment]
ACK & GO

**Comment by smlng** [comment]
ah damn, @geith can you reword you commit message? We typically have the module name first, i.e., `tests: add program for priority inversion problem` would fit here.

**Comment by smlng** [comment]
otherwise we are clear to merge here ...

**Comment by geith** [comment]
done ;-)

**Pull request merged**